src/shape.cpp

```diff
@@ -744,7 +752,7 @@ std::vector<shape> flatten(const std::vector<shape>& shapes)
             auto subs = flatten(s.sub_shapes());
             result.insert(result.end(), subs.begin(), subs.end());
         }
-        else
+        else if(s.type() != shape::uint4_type && s.type() != shape::int4_type)
```

Comment:
This shouldn't be changed, we should still be able to flatten an tuples of int4_type.